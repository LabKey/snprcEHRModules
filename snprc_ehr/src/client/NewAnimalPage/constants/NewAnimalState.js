import moment from 'moment';

const NewAnimalState = () => {

    return  ({
        currentPanel: 1,
        selectedOption: undefined,
        newAnimalData: {
            selectedOption: undefined,
            id: undefined,
            birthDate: { date: moment() },
            bdStatus: undefined,
            acquisitionType: undefined,
            acqDate: { date: moment() },
            gender: undefined,
            sire: undefined,
            dam: undefined,
            species: undefined,
            colony: undefined,
            animalAccount: undefined,
            ownerInstitution: undefined,
            responsibleInstitution: undefined,
            room: undefined,
            cage: { value: undefined },
            diet: undefined,
            pedigree: undefined,
            iacuc: undefined
        },
        summaryData: [],
        speciesList: [],
        acquisitionTypeList: [],
        potentialDamList: [],
        potentialSireList: [],
        locationList: [],
        accountList: [],
        colonyList: [],
        dietList: [],
        institutionList: [],
        iacucList: [],
        pedigreeList: [],
        bdStatusList: [],
        isLoading: true,
        hasError: false,
        preventNext: true,
        saveOk: false,
        showSaveModal: false,
        showCancelModal: false,
        showSpeciesChangeModal: false,
        errorMessage: undefined
    });
}

export default NewAnimalState;
