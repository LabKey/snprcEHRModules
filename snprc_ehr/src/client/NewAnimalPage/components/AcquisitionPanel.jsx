import React from 'react'
import Select from 'react-select'
import moment from 'moment'
import WrappedDatePicker from '../../Shared/components/WrappedDatePicker'
import InfoPanel from '../../Shared/components/InfoPanel'
import { validateNumAnimals, isFutureDate } from '../services/validation'
import constants from '../constants/index'

export default class AcquisitionPanel extends React.Component {
  dateErrorMessageText = 'Future dates are not allowed.'
  state = {
    errorMessage: undefined,
    infoMessage: undefined
  }
componentDidMount = () => {
    this.props.preventNext() // prevent/Allow Next button
  }
handleAcquisitionChange = option => {
    this.props.handleDataChange('acquisitionType', option)
  }
handleAcquisitionDateChange = date => {
  const acqDate = moment(date)
  const dateStatus = isFutureDate(acqDate) || this.props.debug // debug is triggered by test suite
    if (dateStatus) {
      this.props.handleDataChange('acqDate', { date: acqDate })
    }
    this.setState(() => ({
      errorMessage: dateStatus ? undefined : this.dateErrorMessageText
    }))
  }
handleSourceLocationChange = sourceLocation => {
    this.props.handleDataChange('sourceLocation', sourceLocation)
  }
handleNumAnimalChange = e => {
    const numAnimals = e.target.value

    const errorMessage = validateNumAnimals(numAnimals)
    if (errorMessage === undefined) {
      this.props.handleNumAnimalChange(numAnimals)
    }

    this.setState(prevState => (
      {
        ...prevState,
        errorMessage
      }
    ))
  }
isInteger = e => {
    const i = e.key // which ? e.which : e.keyCode;
    const isInteger = (i >= 0 && i <= 9)
    if (!isInteger) {
      e.preventDefault()
    }

    return isInteger
  }
handlePaste = e => {
    e.preventDefault()
  }
render() {
    const { acquisitionType, acqDate, species, sourceLocation, isNonPrimate } = this.props.newAnimalData
    const { numAnimals } = this.props
    const { selectedOption } = this.props

    return (
      <>
        <div className="wizard-panel__rows">
          <div className="wizard-panel__row">
            <div className="wizard-panel__col">
              <WrappedDatePicker
                dateFormat="Pp"
                disabled={ this.props.disabled }
                id="acquisition-datepicker"
                label="Acquisition Date"
                maxDate={ moment().toDate() }
                onChange={ this.handleAcquisitionDateChange }
                onSelect={ this.handleAcquisitionDateChange }
                selected={ acqDate.date.toDate() }
                showTimeSelect
                timeFormat="p"
                timeIntervals={ 30 }
                todayButton="Today"
              />
            </div>
          </div>
          <div className="wizard-panel__row">
            <div className="wizard-panel__col">
              <label className="field-label">Acquisition Code</label>
              <Select
                autoFocus
                className="shared-dropdown"
                classNamePrefix="shared-select"
                id="acquisition-select"
                isClearable
                isDisabled={ this.props.disabled }
                isLoading={ this.props.acquisitionTypeList.length === 0 }
                onChange={ this.handleAcquisitionChange }
                options={ this.props.acquisitionTypeList }
                placeholder="Select Acquisition Type"
                value={ acquisitionType || null }
              />
            </div>
            { selectedOption && selectedOption === 'Acquisition' && (
            <div className="wizard-panel__row">
              <div className="wizard-panel__col">
                <label className="field-label">Source Location</label>
                <Select
                  autoFocus
                  className="shared-dropdown"
                  classNamePrefix="shared-select"
                  id="location-select"
                  isClearable
                  isDisabled={ this.props.disabled }
                  isLoading={ this.props.sourceLocationList.length === 0 }
                  onChange={ this.handleSourceLocationChange }
                  options={ this.props.sourceLocationList }
                  placeholder="Select Source Location"
                  value={ sourceLocation || null }
                />
              </div>
            </div>
            )}
          </div>
          {species && isNonPrimate && (
            <div className="wizard-panel__row">
              <div className="wizard-panel__col">
                <label className="field-label">Number of Animals</label>

                <input
                  className="acq-input"
                  defaultValue={ numAnimals || 1 }
                  disabled={ this.props.disabled }
                  max="100"
                  min="1"
                  onChange={ this.handleNumAnimalChange }
                  onKeyPress={ this.isInteger }
                  onPasteCapture={ this.handlePaste }
                  type="number"
                />

              </div>
            </div>
          ) }
        </div>
        <InfoPanel
          errorMessages={ this.state.errorMessage && (
            [{ propTest: this.state.errorMessage, colName: this.state.errorMessage }]
          ) }
          infoMessages={ species && isNonPrimate && numAnimals && numAnimals !== 1 && [
            { key: 1, value: constants.hamsterWarnings }
          ] }
          messages={
            [{ propTest: !acquisitionType, colName: 'Acquisition Code' }]
          }
        />
      </>
    )
  }
}
