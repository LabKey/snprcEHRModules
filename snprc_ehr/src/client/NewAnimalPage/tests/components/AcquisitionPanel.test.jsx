import React from 'react'
import { shallow } from 'enzyme'
import AcquisitionPanel from '../../components/AcquisitionPanel'
import NewAnimalState from '../../constants/NewAnimalState'
import lists from '../fixtures/lists'

describe('AcquisitionPanel tests', () => {
  test('Should render the AcquisitionPanel', () => {
    const { newAnimalData } = NewAnimalState()

    const wrapper = shallow(
      <AcquisitionPanel
        acquisitionTypeList={ lists.acquisitionTypeList }
        disabled={ false }
        handleDataChange={ () => { } }
        newAnimalData={ newAnimalData }
        preventNext={ () => { return false } }
        numAnimals= { undefined }
        handleNumAnimalChange={ () => {} }
      />

    )
    expect(wrapper).toMatchSnapshot()
  })

  test('Should render the AcquisitionPanel for multiple animal acquisition', () => {
    const newAnimalData = { ...(NewAnimalState().newAnimalData),
                            species: {value: 'HAM', arcSpeciesCode: 'MA' }
    }
    
    const wrapper = shallow(
      <AcquisitionPanel
        acquisitionTypeList={ lists.acquisitionTypeList }
        disabled={ false }
        handleDataChange={ () => { } }
        newAnimalData={ newAnimalData }
        preventNext={ () => { return false } }
        numAnimals={ 2 }
        handleNumAnimalChange={ () => {} }
      />

    )
    expect(wrapper).toMatchSnapshot()
  })
})
