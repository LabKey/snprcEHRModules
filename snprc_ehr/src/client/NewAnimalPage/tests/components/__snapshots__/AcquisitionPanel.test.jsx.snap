// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AcquisitionPanel tests Should render the AcquisitionPanel 1`] = `
<Fragment>
  <div
    className="wizard-panel__rows"
  >
    <div
      className="wizard-panel__row"
    >
      <div
        className="wizard-panel__col"
      >
        <WrappedDatePicker
          dateFormat="Pp"
          disabled={false}
          id="acquisition-datepicker"
          label="Acquisition Date"
          maxDate={1970-01-01T00:00:00.000Z}
          onChange={[Function]}
          onSelect={[Function]}
          selected={1970-01-01T00:00:00.000Z}
          showTimeSelect={true}
          timeFormat="p"
          timeIntervals={30}
          todayButton="Today"
        />
      </div>
    </div>
    <div
      className="wizard-panel__row"
    >
      <div
        className="wizard-panel__col"
      >
        <label
          className="field-label"
        >
          Acquisition Code
        </label>
        <StateManager
          autoFocus={true}
          className="shared-dropdown"
          classNamePrefix="shared-select"
          defaultInputValue=""
          defaultMenuIsOpen={false}
          defaultValue={null}
          id="acquisition-select"
          isClearable={true}
          isDisabled={false}
          isLoading={false}
          onChange={[Function]}
          options={
            Array [
              Object {
                "Category": "Birth",
                "SortOrder": 1,
                "id": 0,
                "label": "1 - Colony-born, Vaginal delivery (at TBRI)",
                "value": "1",
              },
              Object {
                "Category": "Birth",
                "SortOrder": 2,
                "id": 1,
                "label": "2 - Hatched, Colony hatched",
                "value": "2",
              },
            ]
          }
          placeholder="Select Acquisition Type"
          value={null}
        />
      </div>
    </div>
  </div>
  <InfoPanel
    errorMessages={
      Array [
        Object {
          "colName": undefined,
          "propTest": undefined,
        },
      ]
    }
    messages={
      Array [
        Object {
          "colName": "Acquisition Code",
          "propTest": true,
        },
      ]
    }
  />
</Fragment>
`;

exports[`AcquisitionPanel tests Should render the AcquisitionPanel for multiple animal acquisition 1`] = `
<Fragment>
  <div
    className="wizard-panel__rows"
  >
    <div
      className="wizard-panel__row"
    >
      <div
        className="wizard-panel__col"
      >
        <WrappedDatePicker
          dateFormat="Pp"
          disabled={false}
          id="acquisition-datepicker"
          label="Acquisition Date"
          maxDate={1970-01-01T00:00:00.000Z}
          onChange={[Function]}
          onSelect={[Function]}
          selected={1970-01-01T00:00:00.000Z}
          showTimeSelect={true}
          timeFormat="p"
          timeIntervals={30}
          todayButton="Today"
        />
      </div>
    </div>
    <div
      className="wizard-panel__row"
    >
      <div
        className="wizard-panel__col"
      >
        <label
          className="field-label"
        >
          Acquisition Code
        </label>
        <StateManager
          autoFocus={true}
          className="shared-dropdown"
          classNamePrefix="shared-select"
          defaultInputValue=""
          defaultMenuIsOpen={false}
          defaultValue={null}
          id="acquisition-select"
          isClearable={true}
          isDisabled={false}
          isLoading={false}
          onChange={[Function]}
          options={
            Array [
              Object {
                "Category": "Birth",
                "SortOrder": 1,
                "id": 0,
                "label": "1 - Colony-born, Vaginal delivery (at TBRI)",
                "value": "1",
              },
              Object {
                "Category": "Birth",
                "SortOrder": 2,
                "id": 1,
                "label": "2 - Hatched, Colony hatched",
                "value": "2",
              },
            ]
          }
          placeholder="Select Acquisition Type"
          value={null}
        />
      </div>
    </div>
    <div
      className="wizard-panel__row"
    >
      <div
        className="wizard-panel__col"
      >
        <label
          className="field-label"
        >
          Number of Animals
        </label>
        <input
          className="acq-input"
          defaultValue={2}
          disabled={false}
          max="100"
          min="1"
          onChange={[Function]}
          onKeyPress={[Function]}
          onPasteCapture={[Function]}
          type="number"
        />
      </div>
    </div>
  </div>
  <InfoPanel
    errorMessages={
      Array [
        Object {
          "colName": undefined,
          "propTest": undefined,
        },
      ]
    }
    infoMessages={
      Array [
        Object {
          "key": 1,
          "value": "Hamsters can be acquired in bulk; however, when added in bulk you must ensure all animals:
        1) are the same sex
        2) have the same birth and acquisition dates
        3) are in the same location
        4) are on the same account, colony, and IACUC ",
        },
      ]
    }
    messages={
      Array [
        Object {
          "colName": "Acquisition Code",
          "propTest": true,
        },
      ]
    }
  />
</Fragment>
`;

exports[`AcquisitionPanel tests Should render the AcquisitionPanel for multiple animal acquisition 1`] = `
<Fragment>
  <div
    className="wizard-panel__rows"
  >
    <div
      className="wizard-panel__row"
    >
      <div
        className="wizard-panel__col"
      >
        <WrappedDatePicker
          dateFormat="Pp"
          disabled={false}
          id="acquisition-datepicker"
          label="Acquisition Date"
          maxDate={1970-01-01T00:00:00.000Z}
          onChange={[Function]}
          onSelect={[Function]}
          selected={1970-01-01T00:00:00.000Z}
          showTimeSelect={true}
          timeFormat="p"
          timeIntervals={30}
          todayButton="Today"
        />
      </div>
    </div>
    <div
      className="wizard-panel__row"
    >
      <div
        className="wizard-panel__col"
      >
        <label
          className="field-label"
        >
          Acquisition Code
        </label>
        <StateManager
          autoFocus={true}
          className="shared-dropdown"
          classNamePrefix="shared-select"
          defaultInputValue=""
          defaultMenuIsOpen={false}
          defaultValue={null}
          id="acquisition-select"
          isClearable={true}
          isDisabled={false}
          isLoading={false}
          onChange={[Function]}
          options={
            Array [
              Object {
                "Category": "Birth",
                "SortOrder": 1,
                "id": 0,
                "label": "1 - Colony-born, Vaginal delivery (at TBRI)",
                "value": "1",
              },
              Object {
                "Category": "Birth",
                "SortOrder": 2,
                "id": 1,
                "label": "2 - Hatched, Colony hatched",
                "value": "2",
              },
            ]
          }
          placeholder="Select Acquisition Type"
          value={null}
        />
      </div>
    </div>
    <div
      className="wizard-panel__row"
    >
      <div
        className="wizard-panel__col"
      >
        <label
          className="field-label"
        >
          Number of Animals
        </label>
        <input
          className="acq-input"
          defaultValue={2}
          disabled={false}
          max="100"
          min="1"
          onChange={[Function]}
          onKeyPress={[Function]}
          onPasteCapture={[Function]}
          type="number"
        />
      </div>
    </div>
  </div>
  <InfoPanel
    infoMessages={
      Array [
        Object {
          "key": 1,
          "value": "Hamsters can be acquired in bulk; however, when added in bulk you must ensure all animals:
        1) are the same sex
        2) have the same birth and acquisition dates
        3) are in the same location
        4) are on the same account, colony, and IACUC ",
        },
      ]
    }
    messages={
      Array [
        Object {
          "colName": "Acquisition Code",
          "propTest": true,
        },
      ]
    }
  />
</Fragment>
`;
